feat: configure WDK configuration via parsing Cargo manifest metadata

wmmc88 opened the pull request:
This PR enables configuring the WDK Metadata via `[metadata.wdk]` tags in Cargo manifests.

As a consequence of that new feature, several other changes were required:
* Added minimal Sample WDM and UMDF drivers
* Make all driver packaging tooling (ex. signing, cat generation, etc) support WDM and UMDF
* Serde Serializer added to export WDKMetadata configuration to a structured format
* Emit that structured format so that downstream code can `cfg` based off of certain WDK configuration parameters
* A `wdf.c` file is compiled and linked in to provide the `__declspec(selectany)` that are not generated by bindgen
* Have several templates that generate rust code in `wdk-sys`'s build.rs that is otherwise not possible to generate conditionally (ex. call_unsafe_wdf_function_binding definition that points to `OUT_DIR`, `WDFFUNCTIONS` symbol name that depends on configuration, `WDF_FUNCTION_TABLE` definition whose size depends on configuration, etc.)
* added UM support 
* remove local `generated_bindings` copy of bindings
* exporting via `links` keys in cargo manifests is deprecated
* switch to cargo-machete for unused dep detection (required since most code is not cfg-ed out)
* remove wdk-macros running of `cargo check` to obtain out dir (and replace with wdk_sys passing the outdir to the procmacro)

resolves #82

NateD-MSFT commented:
Detailed review is in progress, but quick design question: can you help me understand why we roll a custom (de)serializer rather than building off the TOML crate?

wmmc88 commented:
> Detailed review is in progress, but quick design question: can you help me understand why we roll a custom (de)serializer rather than building off the TOML crate?

@NateD-MSFT the custom serde serializer takes the rust `metadata::Wdk` struct and turns it into a map that's eventually emitted as cargo prints to set `cfg's`. There is no similar "to_map" serializer.

There is no custom deserializer here. `cargo_metadata` [deserializes all the cargo metadata](https://github.com/wmmc88/windows-drivers-rs/blob/b95365c6a5e836f4e5b18da40f46089d1e4a073e/crates/wdk-build/src/lib.rs#L228) into `serde_json::Value`, and then we [deserialize that `serde_json::Value` into the `metadata::Wdk` struct using `serde_json`](https://github.com/wmmc88/windows-drivers-rs/blob/b95365c6a5e836f4e5b18da40f46089d1e4a073e/crates/wdk-build/src/metadata/mod.rs#L188).

wmmc88 commented on crates/wdk-build/src/cargo_make.rs:
```diff
@@ -754,16 +840,106 @@ fn load_wdk_build_makefile<S: AsRef<str> + AsRef<Utf8Path> + AsRef<Path>>(
+/// configurations are detected)
+pub fn package_driver_flow_condition_script() -> anyhow::Result<()> {
+    condition_script(|| {
+        match metadata::Wdk::try_from(&get_cargo_metadata()?) {
```
This doesn't work for workspaces that have WDK configs, but when the current crate being processed is not marked as a driver

wmmc88 replied:
fixed

NateD-MSFT commented on crates/wdk-build/src/lib.rs:
```diff
@@ -605,116 +714,269 @@ impl Config {
     }
 }
 
+/// Find the path the the toplevel Cargo manifest of the currently executing
```
```suggestion
/// Find the path of the toplevel Cargo manifest of the currently executing
```

NateD-MSFT commented on crates/wdk-build/src/lib.rs:
```diff
@@ -605,116 +714,269 @@ impl Config {
+///    1 or 2
+///
+/// The returned path should be a manifest in the same directory of the
+/// lockfile. This does not support invokations that use non-default target
```
```suggestion
/// lockfile. This does not support invocations that use non-default target
```

NateD-MSFT commented on crates/wdk-build/src/lib.rs:
```diff
@@ -605,116 +714,269 @@ impl Config {
+// the future. `EXPORTED_CFG_SETTINGS` is a mapping of cfg key to allowed cfg
+// values
+lazy_static::lazy_static! {
+    // FIXME: replace lazy_static with std::Lazy once available: https://github.com/rust-lang/rust/issues/109736
```
Lazy is now stable if I'm not mistaken.  Not something that needs to be fixed right this instant though.

wmmc88 replied:
> Lazy is now stable if I'm not mistaken. Not something that needs to be fixed right this instant though.

Its available in stable as of 1.80.0 (released 5 days ago). Although not formally stated (or enforced) I've been trying to keep support for Stable and Stable-1. 

There are also other places in the repo where lazy_static is used, that aren't supported by std::sync::Lazy. (ex. LazyLock is not available in no_std, so in those situations, we should migrate to once_cell w/ the `racy` feature)

wmmc88 replied:
tracking this in https://github.com/microsoft/windows-drivers-rs/issues/192

wmmc88 commented on README.md:
````diff
@@ -40,76 +40,77 @@ The crates in this repository are available from [`crates.io`](https://crates.io
+   static GLOBAL_ALLOCATOR: WdkAllocator = WdkAllocator;
    ```
 
    This is only required if you want to be able to use the [`alloc` modules](https://doc.rust-lang.org/alloc/) in the rust standard library. You are also free to use your own implementations of global allocators.
````
At the langauge level, rust has this bring-your-own-allocator construct for no_std programs (ex. km driver). There are still quite a few shortcomings with the `wdk-alloc` crate (ex. panic on oom) so for now I'd like to not force any sort of restrictions here. 

But given that the README is meant as a "getting started", I can change the wording to not explicitly encourage custom allocators

NateD-MSFT replied:
Yeah my concern is longer term - that we may have security or compliance reasons that we require people to use ExAllocatePool* and a specific MS allocator.  But your point about the current allocator being pretty bare bones is perfectly fair.  We can revisit this down the line.

wmmc88 commented on crates/wdk-build/rust-driver-makefile.toml:
```diff
@@ -41,31 +41,79 @@ if is_file ${filepath}
 end
 '''
 
+# This plugin replaces the embedded `CARGO_MAKE_CURRENT_TASK_INITIAL_MAKEFILE_DIRECTORY` in the rust-script doc comment with its evaluated value. See https://github.com/sagiegurari/cargo-make/issues/1081 for more context.
```
Nope. I just didn't want to make the change in this PR since scope is already extremely large here. Will address in a followup. Tracking this in https://github.com/microsoft/windows-drivers-rs/issues/171 already

wmmc88 commented on crates/wdk-build/rust-driver-makefile.toml:
```diff
@@ -283,11 +395,12 @@ wdk_build::cargo_make::copy_to_driver_package_folder(
 private = true
-dependencies = ["copy-sys-to-package", "copy-inf-to-package"]
+dependencies = ["copy-driver-binary-to-package", "copy-inf-to-package"]
+env = { "WDK_BUILD_INF2CAT_OS" = { source = "${CARGO_MAKE_CRATE_TARGET_TRIPLE}", default_value = "UNKNOWN", mapping = { "x86_64-pc-windows-msvc" = "10_x64", "aarch64-pc-windows-msvc" = "Server10_arm64" } } }
```
I used `10_x64` as that was the default value that the WDK resolves to when building C drivers. 

Is there a new value every WDK release (I see `NI` in your suggestion)? If so, there'd need to be some logic here to resolve the mapping here based on the WDK config.

I'm inclined to leave it as is for this PR, since it matches parity with building C drivers

NateD-MSFT replied:
When you say the default value, did you get that from WDK props, or...?

In any event see https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/inf2cat here for the full list of supported OS flags.  You can also provide multiple values to validate for multiple different OSes.  I think at the very least we should go to 10_NI for our default here since Rust drivers are at this early stage experimental and based on our latest OS planning.

wmmc88 replied:
This particular inf2cat setting corresponds with the `Inf2CatWindowsVersionList` property. 

On arm64, it defaults to `Server10_$(DDKPlatform)` where `DDKPlatform` resolves to `arm64`.
On x64, it defaults to `10_$(DDKPlatform)` where `DDKPlatform` resolves to `x64`.
In GE WDK, these defaults are set in `content\Program Files\Windows Kits\10\build\10.0.26100.0\WindowsDriver.OS.Props`.

If you open one of the sample windows drivers (ex. `umdf2toaster.sln`) in VS22, it confirms the above:
**x64:**
![image](https://github.com/user-attachments/assets/77c15645-247f-417d-b16a-bc9634282540)
![image](https://github.com/user-attachments/assets/14308a2d-3715-4201-b911-ce3a98787f66)
**arm64**
![image](https://github.com/user-attachments/assets/ed8fdcb2-29ab-4755-ac2f-cd8a180dd0dc)
![image](https://github.com/user-attachments/assets/e70060f3-5519-4d94-a4dd-30d85be87ad0)

NateD-MSFT replied:
For others reading this review: we resolved this offline and this is not blocking.

wmmc88 commented on crates/wdk-build/src/lib.rs:
```diff
@@ -472,118 +488,203 @@ impl Config {
 
         match &self.driver_config {
-            DriverConfig::WDM() => {
+            DriverConfig::Wdm => {
```
Yes. Ideally we'd be able to consume the WDK's various xml files (vcxproj, targets, etc) and get all of this information. There'd still need to be code here to do things to emit the correctly formatted print statements to pass the information to cargo, but all that information would be parsed from the xmls instead of hardcoded here.

> Maybe I need to build a deserializer for the MSBuild proj files directly
There would need to be work done to be able to merge all the xmls together, since many of the msbuild files reference other xml files and merge them together in some deterministic fashion. I'd imagine merging all that together, and then deserializing that merged xml using some serde xml deserializer(ex. serde_xml_rs) could work.

wmmc88 commented:
Last round of changes for this PR fixed a logic bug that broke automatic skipping of the packaging task for non-binary(ie driver) wdk packages, added more tests for that specific situation, fixed various issues identified by pipelines(ie. formatting, clippy), and expanded coverage of the pipelines to include the workspace-level tests and examples folders.

wmmc88 commented on crates/wdk-build/rust-driver-makefile.toml:
```diff
@@ -391,64 +523,48 @@ run_task = "signtool-verify"
+plugin = "nested-cargo-workspace-in-cargo-make-emulated-workspace-support"
+condition_script_runner_args = [
+  "--base-path",
+  "${CARGO_MAKE_CURRENT_TASK_INITIAL_MAKEFILE_DIRECTORY}",
```
The proper expansion of this is blocked on https://github.com/sagiegurari/cargo-make/pull/1132

wmmc88 replied:
added workaround to unblock pr

wmmc88 commented on crates/wdk-build/rust-driver-makefile.toml:
```diff
@@ -41,28 +44,115 @@ if is_file ${filepath}
+
+    cm_plugin_run_custom_task "{\"run_task\":{\"name\":\"${task.name}\",\"fork\":true}}"
+else
+    # FIXME: modifying the taskjson here is not needed after upstream fix is merged: https://github.com/sagiegurari/cargo-make/pull/1132
```
This should be removed once https://github.com/sagiegurari/cargo-make/pull/1132 is removed

wmmc88 replied:
tracked in #194

wmmc88 commented on crates/wdk-build/src/metadata/error.rs:
```diff
@@ -0,0 +1,57 @@
+    /// error emitted when duplicate key names are found during
+    /// serialization. Serializing into a
+    /// [`metadata::Map`](crate::metadata::Map) requires unique key names
+    #[error(
+        "duplicate keys found during serialization:\nkey: {key}\nvalue 1: {value_1}\nvalue 2: \
+         {value_2}"
+    )]
```
It fails as soon as it finds the first duplicate. In the context of where this sits in the serializer apis and how those apis are invoked, I don't actually think its possible for a duplicate key to be output from the serializer. `-` is used as the seperator between rust identifiers, and rust identifiers aren't allowed to have `-` in them (even raw identifiers can't).

wmmc88 replied:
I had written this as an error instead of a panic because I actually wrote a deserializer as well that could have introduced this as a possible error condition. I ended up removing the deserializer from the scope of this PR, but plan to add it in a future followup

wmmc88 commented on crates/wdk-build/src/cargo_make.rs:
```diff
@@ -754,16 +842,155 @@ fn load_wdk_build_makefile<S: AsRef<str> + AsRef<Utf8Path> + AsRef<Path>>(
+///
+/// # Errors
+///
+/// This function returns an error whenever it determins that the
```
fixed
